wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -198,15 +221,191 @@ def predict(self, X):
+            X_text, max_length=self.max_length, add_special_tokens=True,
+        )
+
+        def gen_train():
+            for i in range(len(X)):
+                features = {k: pad(batch_encoding_text[k][i],
+                                   self.max_length)
+                            for k in batch_encoding_text}
+                features['numerical_features'] = X_numerical[i]
+
+                yield (features, int(y[i]))
+
+        input_element_types = (
+            {**{feature: tf.int32 for feature in text_features},
+             **{'numerical_features': tf.float32}},
+            tf.int64
+        )
+        input_element_tensors = (
+            {**{feature: tf.TensorShape([None]) for feature in text_features},
+             **{'numerical_features': tf.TensorShape([self.n_numerical_features])}},
+            tf.TensorShape([])
+        )
+
+        dataset = tf.data.Dataset.from_generator(
+            gen_train,
+            input_element_types,
+            input_element_tensors,
+        )
+
+        return dataset
+
+    def fit(self, X, y, **kwargs):
+        """
+        Fits semantic classifier
+
+        Args:
+            X(list of t-uples): Assumes that each element x of X has length
+            2+n_numerical_features, the first two components are texts and
+            remaining components are numerical features (e.g.)
+            y: list of 0 (not similar) and 1s (similar)
+            random_state: a random state for the train_test_split
+            epochs: number_of epochs
+
+        Returns:
+             Array of probabilities, of shape len(X) x 2
+
+        """
+        return super().fit(X, y, **kwargs)
+
+    def score(self, X):
+        """
+        Calculates scores for metadata semantic classifier.
+
+        Args:
+            X(list of t-uples): Assumes that each element x of X has length
+            2+n_numerical_features, the first two components are texts and
+            remaining components are numerical features (e.g.)
+
+        Returns:
+            Numpy array of probabilities, of shape len(X) x 2
+
+        """
+
+        X_text, X_numerical = self._separate_features(X)
+
+        X_processed = self._tokenize(X_text)
+        X_processed['numerical_metadata'] = tf.convert_to_tensor(X_numerical)
+
+        predictions = tf.convert_to_tensor(
+            self.meta_model.predict(X_processed)
+        )
+
+        return tf.keras.activations.softmax(predictions).numpy()
+
+    def predict(self, X):
+        """
+        Calculates scores for metadata semantic classifier.
+
+        Args:
+            X(list of t-uples): Assumes that each element x of X has length
+            2+n_numerical_features, the first two components are texts and
+            remaining components are numerical features (e.g.)
+
+        Returns:
+            Array of predictions of length len(X)
+
+        """
+        return super().predict(X)
+
+    def save(self, path):
```

Comment:
isn't the below functions the same for the two models? if all the changes is the name of the variable then i suggest you investigate whether the variable holding the model does not change, same as in sklearn it almost always clf, here it can be model

Reply:
Unfortunately hugging face uses `save_pretrained` while keras use `save` and `load_model` which breaks the abstraction.

I do agree though that calling the main model `self.model` is better consistently and made this change in https://github.com/wellcometrust/WellcomeML/pull/87/commits/e21169377c2d22b766e5da9abe1d91a3641d9f44